yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -218,11 +218,12 @@ private[spark] class ApplicationMaster(
     }
   }
 
-  private def sparkContextInitialized(sc: SparkContext) = {
+  private def sparkContextInitialized(sc: SparkContext, backend: CoarseGrainedSchedulerBackend) = {
     sparkContextRef.synchronized {
       sparkContextRef.compareAndSet(null, sc)
       sparkContextRef.notifyAll()
     }
+    allocator.setScheduler(backend)
```

Comment:
How does this work? This method is called before the `allocator` field is initialized. From `runDriver`, for example:

    // This a bit hacky, but we need to wait until the spark.driver.port property has
    // been set by the Thread executing the user class.
    val sc = waitForSparkContextInitialized()

    // If there is no SparkContext at this point, just fail the app.
    if (sc == null) {
      ...
    } else {
      ....
      registerAM(rpcEnv, sc.ui.map(_.appUIAddress).getOrElse(""), securityMgr) /* this is where allocator is initialized */
      ...
    }